libr/bin/format/elf/elf.c

```diff
@@ -1277,454 +1296,499 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
+
+	return ret;
+}
+
+static void rel_cache_free(HtUPKv *kv) {
+	free (kv->value);
 }
 
+static HtUP *rel_cache_new(ELFOBJ *bin) {
+	RBinElfReloc *relocs = Elf_(r_bin_elf_get_relocs) (bin);
+	const int htsize = R_MIN (bin->reloc_num, 1024);
+	HtUP *rel_cache = ht_up_new_size (htsize, NULL, rel_cache_free, NULL);
+	size_t i;
 
-static ut64 get_import_addr_mips(ELFOBJ *bin, struct ht_rel_t *rel, RBinElfSection *plt_section) {
-	RBinElfSection *s = get_section_by_name (bin, ".rela.plt");
-	if (s) {
-		ut8 buf[1024];
-		const ut8 *base;
-		ut64 plt_addr = s->rva + s->size;
-		int len = r_buf_read_at (bin->b, s->offset + s->size, buf, sizeof (buf));
-		if (len != sizeof (buf)) {
-			// oops
-		}
-		base = r_mem_mem_aligned (buf, sizeof (buf), (const ut8 *)"\x3c\x0f\x00", 3, 4);
-		if (base) {
-			plt_addr += (int)(size_t) (base - buf);
-		} else {
-			plt_addr += MIPS_PLT_OFFSET + 8; // HARDCODED HACK
+	for (i = 0; i < bin->reloc_num; i++) {
+		RBinElfReloc *tmp = R_NEW (RBinElfReloc);
+		memcpy (tmp, relocs + i, sizeof (RBinElfReloc));
+		if (!ht_up_insert (rel_cache, tmp->sym, tmp)) {
+			free (tmp);
 		}
-		plt_addr += rel->k * 16;
-		return plt_addr;
 	}
-	if (plt_section) {
-		const int sizeOfProcedureLinkageTable = 32;
-		const int sizeOfPltEntry = 16;
-		return plt_section->rva + sizeOfProcedureLinkageTable + (rel->k * sizeOfPltEntry);
-	}
-	eprintf ("Unsupported relocs type %" PFMT64u " for arch %d\n",
-		(ut64)REL_TYPE, bin->ehdr.e_machine);
-	return UT64_MAX;
+	free (relocs);
+	return rel_cache;
 }
 
-static ut64 get_import_addr(ELFOBJ *bin, int sym) {
-	int nrel;
-
-	if ((!bin->shdr || !bin->strtab) && !bin->phdr) {
+static ut64 get_got_entrie(ELFOBJ *bin, RBinElfReloc *rel) {
+	if (!rel->rva) {
 		return UT64_MAX;
 	}
 
-	// create rel/rela cache if not already there
-	if (!bin->rel_cache) {
-		bin->rel_cache = rel_cache_new (bin);
-		if (!bin->rel_cache) {
-			return UT64_MAX;
-		}
+	ut64 p_sym_got_addr = Elf_(r_bin_elf_v2p_new) (bin, rel->rva);
+
+	ut8 buf[WORDSIZE];
+	int res = r_buf_read_at (bin->b, p_sym_got_addr, buf, WORDSIZE);
+
+	if (res != WORDSIZE) {
+		return UT64_MAX;
 	}
 
-	// lookup the right rel/rela entry
-	struct ht_rel_t *rel = ht_up_find (bin->rel_cache, sym, NULL);
-	if (!rel) {
+	size_t i = 0;
+	ut64 addr = READWORD (buf, i);
+
+	if (!addr) {
 		return UT64_MAX;
 	}
 
-	RBinElfSection *plt_section = get_section_by_name (bin, ".plt");
-	nrel = bin->rel_cache->count;
+	return addr;
+}
 
-	switch (bin->ehdr.e_machine) {
-	case EM_PPC:
-	case EM_PPC64:
-		return get_import_addr_ppc (bin, rel, plt_section, nrel);
-	case EM_SPARC:
-	case EM_SPARCV9:
-	case EM_SPARC32PLUS:
-		return get_import_addr_sparc (bin, rel, plt_section);
-	case EM_RISCV:
-		return get_import_addr_riscv (bin, rel, plt_section);
+static bool is_thumb_symbol(ut64 plt_addr) {
+	return plt_addr & 1;
+}
+
+static ut64 get_import_addr_arm(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 got_addr = bin->dyn_info.dt_pltgot;
```

Comment:
what if we are working with a ET_REL object (`*.o` files)? They don't have dynamic info, so we should still rely on sections, in some cases, right?

Reply:
Could you had a tests for this case? It will simplify the development for me.

Reply:
I will have a better look later at this, but I may be wrong :)